docs/design/password-key-derivation.md

```diff
@@ -26,51 +26,52 @@ The third requirement allows having multiple encrypted domains to exist within a
+- Because entropy is added as the salt in the `argon2id` step, dictionary attacks on low-entropy passwords cannot be precomputed.
+- The default password can be used for any bucket and will return a distinct key as long as the salt for each bucket differs.
 
 Some other points of consideration include
 
-- Clients can ignore the salt entropy if they wish and know they have a strong password. Indeed, they can ignore all of this and do their own key derivation. At a minimum, we allow the user to specify an already-formed encryption key in their configuration if they don't want one derived from a passphrase.
-- The entropy is not blindly trusted from the satellite, and is mixed with the secret in an HMAC step that ensures a hostile satellite cannot predict what the salt will be.
+- Clients can ignore the salt if they wish and know they have a strong password. Indeed, they can ignore all of this and do their own key derivation. At a minimum, we allow the user to specify an already-formed encryption key in their configuration if they don't want one derived from a passphrase.
+- The salt is not blindly trusted from the satellite, and is mixed with the password in an HMAC step that ensures a hostile satellite cannot predict what the actual salt will be.
 - HMAC provides security against forgeries without knowing the secret, which implies that sharing the output of a HMAC does not leak information about the secret, even under attacker provided data, so I believe the usage of HMAC is safe.
-- The entropy can be used for other purposes in the future if necessary.
-- Users will already lose data if they do not back up their data pointers in the satellite (because the encryption key for the data is only stored there), so adding the bucket entropy does not add a requirement that users need to back up data in a satellite, it just adds a small amount of additional information.
+- Users will already lose data if they do not back up their data pointers in the satellite (because the encryption key for the data is only stored there), so adding the salts does not add a requirement that users need to back up data in a satellite, it just adds a small amount of additional information they need to back up.
 
 ## Implementation
 
-First, satellites must implement the `GetSaltEntropy` call. This can be done by adding a field on buckets and project to store the entropy. It can either run a migration and fill them all at once, or fill it on demand. It should use random bytes from the Go `crypto/rand` package, and be large enough to handle all forseeable usage (32 bytes should be sufficient).
+First, satellites must add a field on buckets and projects to store the salts. It can either run a migration and fill them all at once, or fill it on demand. It should use random bytes from the Go `crypto/rand` package, and be large enough to handle all forseeable usage (32 bytes should be sufficient).
```

Comment:
I'll try and make my case just one more time here 😄: in the case of projects, we already have a guaranteed-unique sequence of bytes associated with each one, and that is the UUID. To add another field as the salt seems to me to suggest that the new field has some different quality from the UUID bytes, and that said difference is important for our security or operating costs. But I think it can be demonstrated that that is not the case: the UUID and the proposed salt have exactly the same exposure and accessibility (both entirely public); the UUID has equivalent or stronger requirements in terms of the amount of entropy and the randomness quality; the lifetimes are identical and neither has any need to allow change; and the failure domain when either field is generated incorrectly or compromised is not meaningfully different. To add a new field containing additional entropy beyond the amount which is useful just adds more load on the db with no benefit.

My preference for the bucket salts to be referred to as "bucket IDs" or to be treated as actual UUIDs involves a similar argument, but it doesn't carry the same cost of db load, so it doesn't matter as much.

Reply:
Using the project UUID as the project salt will make the implementation easier too, as no changes would be required on the satellite.

Reply:
So, for buckets, either way we need to add something to bucket metadata - either a UUID or a salt. Buckets in particular though don't need a UUID, the bucket name within a project is sufficient, so adding a UUID feels unnecessary. We do need a salt for buckets, but we don't need a UUID. Right?

For projects, if we use the project UUID as the project salt, then we don't need to add project metadata, but a previously unencumbered data type now has important consequences cryptographically. Further, I'm not convinced yet about how project UUIDs will work with Satellite import/export. Basically, I'm not actually sure projects need UUIDs. I get that they're useful now, but I'm not sure that will always be the case. It might be, but I don't know.

Separating these concerns to me does feel like the right move. 

One additional caveat - we're really hoping this design doc can be *implemented* before the release this sprint, which is... Tuesday. So perhaps we can implement the bucket metadata first and punt on the project metadata anyways?

Let's do this - let's use bucket salts (we don't need bucket UUIDs). For project salts, let's use the project UUID for now but abstract it in the API call so we can change it to generic stable project entropy, should we ever change while designing import/export to use something to identify projects other than UUIDs. And then if we can do this in such a way that we can implement the basic things ASAP, that would be great!

Reply:
So, bucket salts and project UUIDs with the caveat that the design refers to it as a salt throughout? IOW, the project salt will be backed by the project UUID.

Reply:
that is my proposal

Reply:
> adding a UUID feels unnecessary

I'm only saying that we're already essentially adding a unique identifier, in the form of the bytes to use for salting. Why not make it official? Or is the concern that a UUID is more entropy than needed?

> Bucket names are by design unique within a project and unlikely to change, both within our pointer data model and by product requirements in regards to S3 similarity.

I think the motivating factor for a lot of the design here is being able to rename buckets. If we don't care about that, I would suggest that we use the content-key derivation algorithms already in place, and limit this doc to deriving "root" encryption keys from human-supplied passphrases.

I'd rather be able to rename buckets, though, if it's all the same.

> if we use the project UUID as the project salt, then it's true that we don't need to add project metadata, but a previously unencumbered data type now has important consequences cryptographically.

The _only_ quality with any cryptographic significance whatsoever here is uniqueness. As long as salts are unique, or even _mostly_ unique, the goal of foiling rainbow tables is accomplished. It could be dangerous to depend on the entropy from a salt in any respect beyond that.

Reply:
> I think the motivating factor for a lot of the design here is being able to rename buckets. 

I disagree. The only part of the design that changes if we don't care about renaming buckets is we can remove the salt per bucket.

---

I can imagine a world where a user can request a new salt or even store their own salt with the satellite. I can also imagine a world where we regenerate UUIDs for things (import/export into another satellite). To me, a UUID communicates different intent and properties than salt does, and it seems dangerous to mix them, even if one can be used as the other. For example, why not use project creation time as the salt? It has the same properties as the UUID, but that'd be super confusing.

Reply:
* this discussion need not block merging of the PR.

> I disagree. The only part of the design that changes if we don't care about renaming buckets is we can remove the salt per bucket.

That's true. Project salts would still be worthwhile. You're right.

> I can imagine a world where a user can request a new salt or even store their own salt with the satellite.

My imagination is not up to the task here. Could you share a scenario so we can be on the same page about what we want to provide?

> I can also imagine a world where we regenerate UUIDs for things (import/export into another satellite).

I can easily believe that there's a good reason but I'm not seeing it. Why wouldn't we use the same UUID on the other satellite? Or under what conditions would we need to regenerate a UUID?

> To me, a UUID communicates different intent and properties than salt does, and it seems dangerous to mix them, even if one can be used as the other.

We may not even be on the same page about what UUIDs and salts are. To me they both signify the intent and properties of uniqueness, nothing more or less, but with possibly varying degrees of surety that they're _actually_ unique. If you're imputing any more qualities than that to a salt, that might be a recipe for trouble.

> For example, why not use project creation time as the salt? It has the same properties as the UUID (for the purposes of being a salt), but that'd be super confusing.

If project creation time was expected to be universally unique (across multiple satellites), then yes, it might well be worth using that value for both the salt and for project identification in place of a UUID. But I don't think that's a quality we have in our creation times, so I don't think it would make a good salt.

Reply:
> My imagination is not up to the task here. Could you share a scenario so we can be on the same page about what we want to provide?

This design goes through some effort to ensure satellites controlling salts can't be used to gain information about users passwords or keys. Maybe very paranoid users that don't trust my security understanding wish to use their own salts. Maybe we eventually don't trust my security understanding and enforce that clients store salts that they generate with the servers.

> I can easily believe that there's a good reason but I'm not seeing it. Why wouldn't we use the same UUID on the other satellite? Or under what conditions would we need to regenerate a UUID?

If the UUID collides (lol). Maybe a user wants to import the same project twice in their own satellite? Maybe a backup/shadow copy?

> We may not even be on the same page about what UUIDs and salts are. To me they both signify the intent and properties of uniqueness, nothing more or less, but with possibly varying degrees of surety that they're actually unique. If you're imputing any more qualities than that to a salt, that might be a recipe for trouble.

To me, a UUID is an identifier. Indeed, it contains enough entropy to be universally unique, but what it **is** is an identifier to something. Salts do not need to be universally unique, nor is it much of a failure if they aren't totally unique even in the same satellite. To me, they're just a source of entropy to make creating tables of weak keys impossible.

In other words, to me, a UUID conveys information that there's something that can be looked up if I have it, where a salt does not. So it's not a problem of a UUID being unfit, it's a problem of it providing too much.

> If project creation time was expected to be universally unique (across multiple satellites), then yes, it might well be worth using that value for both the salt and for project identification in place of a UUID. But I don't think that's a quality we have in our creation times, so I don't think it would make a good salt.

I agree creation time may not be unique, and so I think my point was lost due to our different perspectives. If I found a database that existed with a uniqueness constraint on creation time and found that the creation time was being used as a salt, I'd find that to be a massive subversion of my expectations. 

Sure, it may save a field, but why not just have it stored distinctly like how every other system does it? In my experience, salts are stored as part of a blob containing it and the password hash (and perhaps some data about the algorithm), but that's logically the same thing as it being a separate field. The only difference our system has is the password hashes are stored on the client and the salts are stored on the server. I don't think I've ever seen a system reuse a field for the purpose of being a salt.

Reply:
(I'm carrying on with this, even though the PR is approved, just with the hopes of getting to where I feel like we understand each other. Ignore with impunity if desired.)

> This design goes through some effort to ensure satellites controlling salts can't be used to gain information about users passwords or keys. Maybe very paranoid users that don't trust my security understanding wish to use their own salts. Maybe we eventually don't trust my security understanding and enforce that clients store salts that they generate with the servers.

So, the only thing afforded by salt in our construction is that it makes rainbow tables much less effective. There isn't anything there to trust. They don't even need to trust that their salt is unique, because they can inspect the salts of other projects/buckets. If they think we can subvert their security with a specially-crafted salt, then they don't trust SHA256 or Argon2 or AES128-GCM (either because they don't trust crypto in general, or because they know about some extremely alarming weaknesses in these that no one else knows about). I don't think we can satisfy such users no matter what we do.

> If the UUID collides (lol). Maybe a user wants to import the same project twice in their own satellite? Like as a backup/shadow copy?

That sounds like it would multiply our metadata costs without the user paying for any extra space, and that would be bad. If they do want to store everything again, then fantastic--and they'll want a new UUID for the backup project.

And if we experience unexpected UUID collisions before having something like 10^15 projects, that is an event that we want to cause sirens and flashing red lights.

> Sure, it may save a field, but why not just have it stored distinctly like how every other system does it? In my experience, salts are stored as part of a blob containing it and the password hash (and perhaps some data about the algorithm), but that's logically the same thing as it being a separate field. The only difference our system has is the password hashes are stored on the client and the salts are stored on the server. I don't think I've ever seen a system reuse a field for the purpose of being a salt.

Maybe not when the field is _called_ a "salt". But using some unique property of an item to modulate a hash or an encryption is so common that I'd be surprised to see any modern crypto-using system that _doesn't_ do it somewhere. Every time someone does an HMAC with a username or a userID or a URL or an email address, they're doing what I am proposing. Perhaps the disconnect comes about here because people only tend to call such a modulator a "salt" when they don't have any _other_ unique properties they can use and they have to store excess entropy for this sole purpose.

Reply:
I think we should continue this conversation and I'm enjoying it, but

1. I don't think I can prioritize the deep thinking right now
2. It'd probably go 10x faster in person.

Reply:
Well stated.

I’m satisfied that if we ever do need to separate the concerns of project uuid and salt, at least it can be made nicely backwards compatible.